pkg/spi/spi.go

```diff
@@ -16,49 +16,126 @@
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	spiapi "github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	cdqanalysis "github.com/redhat-appstudio/application-service/cdq-analysis/pkg"
 	"github.com/redhat-appstudio/application-service/pkg/devfile"
-	"github.com/redhat-appstudio/service-provider-integration-scm-file-retriever/gitfile"
 )
 
 type SPI interface {
-	GetFileContents(ctx context.Context, namespace string, repoUrl string, filepath string, ref string, callback func(ctx context.Context, url string)) (io.ReadCloser, error)
+	GetFileContents(ctx context.Context, name string, component v1alpha1.Component, repoUrl string, filepath string, ref string) (io.ReadCloser, error)
 }
 
+const (
+	SPIFCR_waiting_for_delivered_phase = "File content request status has not been delivered"
+	SPIFCR_prefix                      = "spi-fcr-"
+)
+
 type SPIClient struct {
+	K8sClient client.Client
+}
+
+// SPIFileContentRequestError returns an internal error
+type SPIFileContentRequestError struct {
+	Message string
+}
+
+func (e *SPIFileContentRequestError) Error() string {
+	return "SPIFileContentRequest failed " + e.Message
 }
 
+var ValidDevfileLocations = []string{cdqanalysis.Devfile, cdqanalysis.HiddenDevfile, cdqanalysis.HiddenDirDevfile, cdqanalysis.HiddenDirHiddenDevfile}
+
 // GetFileContents is a wrapper call to scm file retriever's GetFileContents()
-func (s SPIClient) GetFileContents(ctx context.Context, namespace string, repoUrl string, filepath string, ref string, callback func(ctx context.Context, url string)) (io.ReadCloser, error) {
-	return gitfile.Default().GetFileContents(ctx, namespace, repoUrl, filepath, ref, callback)
+func (s SPIClient) GetFileContents(ctx context.Context, name string, component v1alpha1.Component, repoUrl string, filepath string, ref string) (io.ReadCloser, error) {
+	log := ctrl.LoggerFrom(ctx)
+	spiFCRLookupKey := types.NamespacedName{Name: SPIFCR_prefix + name, Namespace: component.Namespace} //We'll use a unique component name and filepath to construct the name
+	spiFCR := &spiapi.SPIFileContentRequest{}
+	log.Info("Looking up SPIFileContentRequest CR", "name", spiFCRLookupKey.Name, "namespace", spiFCRLookupKey.Namespace)
+	err := s.K8sClient.Get(ctx, spiFCRLookupKey, spiFCR)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			spiFCR.Name = spiFCRLookupKey.Name
+			spiFCR.Namespace = spiFCRLookupKey.Namespace
+			spiFCR.Spec.RepoUrl = repoUrl
+			spiFCR.Spec.FilePath = filepath
+			spiFCR.Spec.Ref = ref
+			//add an owner reference
+			ownerReference := metav1.OwnerReference{
+				APIVersion: component.APIVersion,
+				Kind:       component.Kind,
+				Name:       component.Name,
+				UID:        component.UID,
+			}
+			spiFCR.SetOwnerReferences(append(spiFCR.GetOwnerReferences(), ownerReference))
```

Comment:
do they just hang around until the component is deleted?

Reply:
yes, or the SPI controller deletes them after 30 mins